### PR TITLE
Add missing parts for provisioning

### PR DESCRIPTION
**Pull request by ChriMarMe**
Signed-off-by: Christopher Meis <christopher.meis@9elements.com>

### CODE REVIEW

File: cmd/txt-prov/cmd.go [hunk 1]
```diff
@@ -24,6 +26,9 @@ var (
 	savePol       = flag.Bool("s", false, "Select if policies shall be saved in a file. Default: false")
 	help          = flag.Bool("h", false, "Shows help")
 	version       = flag.Bool("v", false, "Shows version and license information")
+
+	// ProvisioningConfigFileStdName
+	lcpfilejson = "provdataTPM2.json"
```
Comment (zaolin) [inline cmd/txt-prov/cmd.go:31]: parse filename from command line agrs

File: cmd/txt-prov/cmd.go [hunk 2]
```diff
@@ -120,3 +125,64 @@ func loadFiles() ([]byte, []byte, []byte, *crypto.Hash, error) {
 	}
 	return delHash, writeHash, psPolicy, hAlg, nil
 }
+
+func getLCPDataFromFile() (*tools.LCPPolicy2, error) {
+	var ok bool
+	var lcppol tools.LCPPolicy2
+	var b []byte
+	f, err := os.Open(lcpfilejson)
```
Comment (zaolin) [inline cmd/txt-prov/cmd.go:133]: ioutil.ReadFile

File: cmd/txt-prov/cmd.go [hunk 3]
```diff
@@ -120,3 +123,60 @@ func loadFiles() ([]byte, []byte, []byte, *crypto.Hash, error) {
...
+		lcppol.Version = lcpdata.Version
+	}
+
+	lcppol.HashAlg, ok = tools.HashAlgMap[prov.HashMapping[lcpdata.HashAlg]]
+	if ok == false {
+		return nil, fmt.Errorf("Cant determin hash algorithm")
+	}
+
+	if (lcpdata.PolicyType != uint8(0)) && (lcpdata.PolicyType != uint8(1)) {
+		lcppol.PolicyType = tools.LCPPolicyTypeAny
```
Comment (PatrickRudolph) [inline cmd/txt-prov/cmd.go:149]: return error

File: cmd/txt-prov/cmd.go [hunk 4]
```diff
@@ -120,3 +123,60 @@ func loadFiles() ([]byte, []byte, []byte, *crypto.Hash, error) {
...
+	var lcppol tools.LCPPolicy2
+	var b []byte
+	b, err := ioutil.ReadFile(*pslcpfile)
+	if err != nil {
+		return nil, err
+	}
+	lcpdata, err := parseLCP2File(b)
+	// Parse lcpdata field
+	if lcpdata.Version < uint16(0x300) || lcpdata.Version > uint16(0x306) {
+		lcppol.Version = uint16(0x302)
```
Comment (PatrickRudolph) [inline cmd/txt-prov/cmd.go:138]: return error

File: cmd/txt-prov/cmd.go [hunk 5]
```diff
@@ -120,3 +123,60 @@ func loadFiles() ([]byte, []byte, []byte, *crypto.Hash, error) {
...
+func getLCPDataFromFile() (*tools.LCPPolicy2, error) {
+	var ok bool
+	var lcppol tools.LCPPolicy2
+	var b []byte
+	b, err := ioutil.ReadFile(*pslcpfile)
+	if err != nil {
+		return nil, err
+	}
+	lcpdata, err := parseLCP2File(b)
+	// Parse lcpdata field
```
Comment (PatrickRudolph) [inline cmd/txt-prov/cmd.go:136]: check err is not nil

File: pkg/provisioning/psprov.go [hunk 6]
```diff
@@ -41,10 +45,16 @@ func ProvisionTPM20(rw io.ReadWriter, delHash, writeHash []byte, hashAlg crypto.
...
 
 func definePSIndexTPM20(rw io.ReadWriter, pspolhash []byte, hashAlg crypto.Hash) error {
+	hwAPI := hwapi.GetAPI()
 	tpm2PSIndexDef.AuthPolicy = pspolhash
 	tpm2PSIndexDef.NameAlg = tpm2.Algorithm(hashAlg)
 
 	err := tpm2.NVDefineSpaceEx(tpmCon.RWC, tpm2.HandleOwner, emptyPW, emptyPW, tpm2PSIndexDef)
+	for _, item := range tests.TestsTPM {
+		if strings.Contains(item.Name, "PS Index has correct config") {
+			item.Run(hwAPI)
```
Comment (PatrickRudolph) [inline pkg/provisioning/psprov.go:55]: check error value or return value

File: pkg/provisioning/psprov.go [hunk 7]
```diff
@@ -168,3 +178,45 @@ func mergeToPSPolicy(rw io.ReadWriter, delPol, writePol []byte) ([]byte, error)
...
+	if err != nil {
+		return err
+	}
+	zerohashraw, err := tpmutil.Pack(zeroHash)
+	if err != nil {
+		return err
+	}
+	writehashdata := tpmutil.U16Bytes(writehashraw)
+	zeorhashdata := tpmutil.U16Bytes(zerohashraw)
+	err = tpm2.PolicyOr(rw, sess, tpm2.TPMLDigest{Count: uint32(2), Digests: []tpmutil.U16Bytes{writehashdata, zeorhashdata}})
```
Comment (PatrickRudolph) [inline pkg/provisioning/psprov.go:204]: Or(hash, Zero) == hash.
Why use OR at all?

Comment (ChriMarMe) [reply pkg/provisioning/psprov.go:198]: TPM2 Spec, Part 3, Page 227,  23.6 TPM2_PolicyOr

File: cmd/txt-prov/cmd.go [hunk 8]
```diff
@@ -64,59 +72,128 @@ func handlePasswordsTPM12() ([20]byte, [20]byte) {
...
+	var b []byte
+	b, err := ioutil.ReadFile(*pslcpfile)
+	if err != nil {
+		return nil, err
+	}
+	lcpdata, err := parseLCP2File(b)
+	if err != nil {
+		return nil, err
+	}
+	ver, err := strconv.ParseUint(lcpdata.Version, 0, 16)
```
Comment (zaolin) [inline cmd/txt-prov/cmd.go:127]: strconv.ParseUint(lcpdata.Version, 0, 0)

File: cmd/txt-prov/cmd.go [hunk 9]
```diff
@@ -64,59 +72,128 @@ func handlePasswordsTPM12() ([20]byte, [20]byte) {
...
+	if uint16(ver) < uint16(0x300) || uint16(ver) > uint16(0x306) {
+		return nil, fmt.Errorf("Invalid LCP Version. Want: 0x302 - 0x306 - Have: %v", lcpdata.Version)
+	}
+
+	hashAlg, ok := tools.HashAlgMap[prov.HashMapping[lcpdata.HashAlg]]
+	if ok == false {
+		return nil, fmt.Errorf("Cant determin hash algorithm")
+	}
+
+	pT, err := strconv.ParseUint(lcpdata.PolicyType, 10, 8)
```
Comment (zaolin) [inline cmd/txt-prov/cmd.go:140]: pT, err := strconv.ParseUint(lcpdata.PolicyType, 0, 0)

File: cmd/txt-prov/cmd.go [hunk 10]
```diff
@@ -64,59 +72,128 @@ func handlePasswordsTPM12() ([20]byte, [20]byte) {
...
+	if err != nil {
+		return nil, err
+	}
+	if tools.LCPPolicyType(pT) > tools.LCPPolicyTypeAny && tools.LCPPolicyType(pT) < tools.LCPPolicyTypeList {
+		return nil, fmt.Errorf("Invalid PolicyType. Want: 0(Signed Policy) or 1 (Auto promotion) - Have: %v", lcpdata.PolicyType)
+	}
+
+	smv, err := strconv.ParseUint(lcpdata.SINITMinVersion, 10, 8)
```
Comment (zaolin) [inline cmd/txt-prov/cmd.go:148]: smv, err := strconv.ParseUint(lcpdata.SINITMinVersion, 0, 0)

File: cmd/txt-prov/cmd.go (hunk 9 above)
Comment (zaolin) [reply cmd/txt-prov/cmd.go:140]: This will probably fix your issue

**Review by zaolin — APPROVED** [review]

**Pull request merged**